libwacom/libwacom-database.c

```diff
@@ -295,32 +412,71 @@ libwacom_parse_stylus_keyfile(WacomDeviceDatabase *db, const char *path)
 
 	groups = g_key_file_get_groups (keyfile, NULL);
 	for (i = 0; groups[i]; i++) {
-		WacomStylus *stylus;
+		WacomStylus *stylus = NULL, *aliased = NULL;
 		WacomStylusId id;
 		GError *error = NULL;
 		char *eraser_type, *type;
 		char **string_list;
+		int aliased_int;
 
 		if (!parse_stylus_id(groups[i], &id)) {
 			g_warning ("Failed to parse stylus ID '%s', ignoring entry", groups[i]);
 			continue;
 		}
+
+		char *aliasstr = g_key_file_get_string(keyfile, groups[i], "AliasOf", NULL);
+
+		if (handle_aliases == IGNORE_ALIASES && aliasstr) {
+			g_clear_pointer(&aliasstr, g_free);
+			continue;
+		}
+
+		if (handle_aliases == ONLY_ALIASES && !aliasstr) {
+			g_clear_pointer(&aliasstr, g_free);
+			continue;
+		}
+
 		stylus = g_new0 (WacomStylus, 1);
 		stylus->refcnt = 1;
 		stylus->id = id;
-		stylus->name = g_key_file_get_string(keyfile, groups[i], "Name", NULL);
-		stylus->group = g_key_file_get_string(keyfile, groups[i], "Group", NULL);
+
+		if (handle_aliases == ONLY_ALIASES) {
+			WacomStylusId lookup;
+
+			if (!parse_stylus_id(groups[i], &lookup)) {
+				g_warning ("Failed to parse Aliased ID '%s', ignoring entry", groups[i]);
+				continue;
+			}
+			if (safe_atoi_base (aliasstr, &aliased_int, 16))
+				lookup.tool_id = aliased_int;
+
+			aliased = g_hash_table_lookup(db->stylus_ht, &lookup);
+			if (!aliased) {
+				g_warning ("Unkown Aliased ID 0x%08x %s\n", lookup.tool_id, aliasstr);
+				continue;
```

Comment:
both of the `continue`s here will leak `stylus`.

You could trigger this with a crafted `StylusEntry` in `test_libwacom.py` but since we don't run those tests in valgrind.... :shrug: